### ISSUE TITLE
Is there any trick to boost EvalHook?

### ISSUE THREAD

**Issue by HarborYuan** [open-mmlab/mmcv#1348]
Hi,

Thanks for your work on mmcv.

Currently, I am wokring on my own project with mmseg and mmdet. But I found that EvalHook performs slower than I need.
Based on my understanding, the current version of DistEvalHook will excute in the following process:

>>>  Model Inference -> gpu/cpu collect -> dataset.evaluate()

With cpu, the dataset.evaluate is too slow in my project, while with gpu(custom metirc model written with nn.Module built in the dataset class), the GPU memory comsumption will be too large. 

My question is how to write a custom EvalHook to boost the whole process. For example, can I do something to dump the model to RAM or do the dataset.evaluate() with cpu and let the runner go ahead to train the model with the following data at the same time.

As a fan of the openmm-lab projects, I would be really appreciate if there could be some features to boost the EvalHook in the future version of mmcv/mmdet/mmseg. I would be happy to contribute to this if possible.

Thanks.

**Comment by zhouzaida** [comment]
hi @HarborYuan , sorry for the late reply. Did you mean `dataset.evaluate()` and training the next epoch are excuting at the same time?

**Comment by HarborYuan** [comment]
@zhouzaida Thanks for your response. That is what I mean.

**Comment by zhouzaida** [comment]
Can you provide the ratio of the time it consumes to the evaluating time (`Time of dataset.evaluate() / Time of evaluating`)?

**Comment by HarborYuan** [comment]
I will report this in the next few days.
For now, what I know is that it seems that each one of the following: 1.data loader preparation of validation set, 2.cpu collect(involve disk I/O), 3.dataset.evaluation may take considerable time compared to gpu inference. I give dataset.evaluation a progress bar and it will perform multiple times slower than the gpu inference bar.

**Comment by HarborYuan** [comment]
In my project, it will take ~198 seconds to execute dataset.evaluate() but only need ~30 seconds to the inference process, although the evaluate codes are not complicated.
It will also take ~85s to cpu collect the data, this may be due to the I/O with disk. 

GPU evalute will diminish the dataset.evalute drastically and even lower than the inference time, but will double the GPU RAM cosumption.

In conclusion, with hot cache, dataset.evaluate() / Time of evaluating is ~0.63. Cold cache may lead to longer total time.

**Comment by zhouzaida** [comment]
Currently, mmcv can not support executing `dataset.evaluate()` and training the next epoch in parallel. Do you have any suggestions?

**Comment by HarborYuan** [comment]
Is it possible to refer to the [metric](https://github.com/PyTorchLightning/metrics) project's design, and build the mmcv's evaluation class to support distributed evaluation with GPU?

**Comment by zhouzaida** [comment]
In fact, we also support distributed evaluation.
https://github.com/open-mmlab/mmcv/blob/9d4571e3d0bddb7e49852909eddcf5b4145c1a91/mmcv/runner/hooks/evaluation.py#L420-L450

However, it temporarily does not support calculating the key_score and training the next epoch in parallel.

https://github.com/open-mmlab/mmcv/blob/9d4571e3d0bddb7e49852909eddcf5b4145c1a91/mmcv/runner/hooks/evaluation.py#L447

https://github.com/open-mmlab/mmcv/blob/9d4571e3d0bddb7e49852909eddcf5b4145c1a91/mmcv/runner/hooks/evaluation.py#L317-L336

**Comment by zhouzaida** [comment]
The issue is a nice question and we need more discussion.

**Comment by HarborYuan** [comment]
Thanks for your response. Based on my understanding, to support distributed evaluation on multiple gpus, maybe it could be better if we can define a module contains "state" which stores the Intermediate variables. The "state" can be updated after each inference iter, and collected finally with custom strategy (e.g., mean / add / concat).

I guess this abstraction may support most of the evaluation algorithms for now. Compared to the current "Distributed inference -> collect results -> evaluation" strategy, it may boost the eval hook drastically.

**Comment by zhouzaida** [comment]
The PR (https://github.com/open-mmlab/mmsegmentation/pull/709) has supported the feature that calculating the intermediate results to reduce the consumption of memory.
